File: 953 - Verifying an Alien Dictionary/cpp/ans.cpp
class Solution {
public:
    // method 1, using hashmap
    bool isAlienSorted(vector<string>& words, string order) {
        // convert the order to hash map
        unordered_map<char, int> hmap;
        for (int i = 0; i < order.length(); i++) {
            hmap[order[i]] = i;
        }
        // loop through the list
        for (int i = 0; i < words.size() - 1; i++) {
            string w1 = words[i];
            string w2 = words[i+1];
            // check every character in w1 and w2
            for (int j = 0; j < w1.length(); j++) {
                if (j == w2.length()) {
                    // reaches the end of w2, that means w2 is a prefix of w1
                    return false;
                }
                if (w1[j] != w2[j]) {
                    // first different char
                    // check the value using the hashmap
                    if (hmap[w1[j]] > hmap[w2[j]]) {
                        // the current char in w1 comes after char in w2
                        return false;
                    }
                    // stop checking the current words, and start the next one 
				    break;
                }
            }
        }
        return true;
    }
    // method 2, using find
    bool isAlienSorted(vector<string>& words, string order) {
        // loop through the list
        for (int i = 0; i < words.size() - 1; i++) {
            string w1 = words[i];
            string w2 = words[i+1];
            // check every character in w1 and w2
            for (int j = 0; j < w1.length(); j++) {
                if (j == w2.length()) {
                    // reaches the end of w2, that means w2 is a prefix of w1
                    return false;
                }
                if (w1[j] != w2[j]) {
                    // first different char
                    // check the value using the hashmap
                    if (order.find(w1[j]) > order.find(w2[j])) {
                        // the current char in w1 comes after char in w2
                        return false;
                    }
                    // stop checking the current words, and start the next one 
				    break;
                }
            }
        }
        return true;
    }
};

File: 953 - Verifying an Alien Dictionary/rust/ans.rs
use std::collections::HashMap;

impl Solution {
    // method 1
    pub fn is_alien_sorted(words: Vec<String>, order: String) -> bool {
        // store the order to hash map
        let mut hmap = HashMap::new();
        for (i, x) in order.as_bytes().iter().enumerate() {
            hmap.insert(x, i);
        }
        // loop through the list
        for i in 0..words.len()-1 {
            let w1 = words[i].as_bytes();
            let w2 = words[i + 1].as_bytes();
            // check every char in words
            for j in 0..w1.len() {
                if j == w2.len() {
                    // reaches the end of w2, means w2 is a prefix of w1
                    return false;
                }
                if w1[j] != w2[j] {
                    // find the first difference
                    // check the value in hashmap
                    if hmap.get(&w1[j]) > hmap.get(&w2[j]) {
                        // the current char in w1 comes after char in w2
                        return false;
                    }
                    // stop checking the current words, and start the next one 
				    break;
                }
            }
        }

        return true;
    }
    // method 2
    pub fn is_alien_sorted(words: Vec<String>, order: String) -> bool {
        // loop through the list
        for i in 0..words.len()-1 {
            let w1 = words[i].as_bytes();
            let w2 = words[i + 1].as_bytes();
            // check every char in words
            for j in 0..w1.len() {
                if j == w2.len() {
                    // reaches the end of w2, means w2 is a prefix of w1
                    return false;
                }
                if w1[j] != w2[j] {
                    // find the first difference
                    // check the value in hashmap
                    if order.find(w1[j] as char) > order.find(w2[j] as char) {
                        // the current char in w1 comes after char in w2
                        return false;
                    }
                    // stop checking the current words, and start the next one 
				    break;
                }
            }
        }

        return true;
    }
}